torch/_functorch/_activation_checkpointing/ac_reorder_inference_pass.py

```diff
@@ -0,0 +1,168 @@
+    4. DCE cleans up unused forward copies
+
+    Args:
+        gm: Graph with forward and backward ops (inference mode joint graph)
+
+    Returns:
+        Graph with rematerialized checkpoint regions
+    """
+    # Find backward boundary
+    first_node_in_bwd = None
+    for node in gm.graph.nodes:
+        if _is_backward_node(node):
+            first_node_in_bwd = node
+            break
+
+    if first_node_in_bwd is None:
+        return gm
+
+    # Build ordering
+    order = {node: idx for idx, node in enumerate(gm.graph.nodes)}
+    bwd_start = order[first_node_in_bwd]
+
+    # Track which checkpointed nodes are used in backward
+    checkpointed_in_bwd: set[fx.Node] = set()
+    for node in gm.graph.nodes:
+        if must_recompute(node):
+            for user in node.users:
+                if order[user] >= bwd_start:
+                    checkpointed_in_bwd.add(node)
+                    break
+
+    # Build reordered graph
+    new_graph = fx.Graph()
+    env: dict[fx.Node, fx.Node] = {}
+    recomputed_nodes: dict[fx.Node, fx.Node] = {}
+
+    # Add placeholders (like reordering_to_mimic_autograd_engine)
+    for node in gm.graph.find_nodes(op="placeholder"):
+        env[node] = new_graph.node_copy(node, lambda x: env[x])
+
+    def duplicate_checkpoint_chain(node: fx.Node) -> fx.Node:
+        """
+        Duplicate a checkpointed node and its checkpointed dependencies.
+        Stops at non-checkpointed nodes (checkpoint inputs) - reuses those from forward.
+        """
+        if node in recomputed_nodes:
+            return recomputed_nodes[node]
+
+        # Recursively duplicate checkpointed dependencies first
+        for inp in node.all_input_nodes:
+            if inp.op == "placeholder":
+                continue
+            if must_recompute(inp) and inp not in recomputed_nodes:
+                duplicate_checkpoint_chain(inp)
+
+        # Create duplicate node
+        def get_input(x: fx.Node) -> fx.Node:
+            # Use recomputed version if available
+            if x in recomputed_nodes:
+                return recomputed_nodes[x]
+            # Otherwise reuse from forward (checkpoint inputs)
+            return env.get(x, x)
+
+        dup = new_graph.node_copy(node, get_input)
+        dup.name = node.name + "_recomputed"
+        recomputed_nodes[node] = dup
+        return dup
+
+    def insert_node_in_graph(node: fx.Node, for_backward: bool) -> None:
+        """
+        Insert a node and its dependencies into the graph.
+        Pattern from reordering_to_mimic_autograd_engine.
+        """
+
+        # Skip condition: already in env (but in backward, may need to duplicate AC nodes)
+        def skip_condition(n: fx.Node) -> bool:
+            if n in recomputed_nodes:
+                return True
+            if n in env:
+                # In backward, we need to duplicate checkpointed nodes even if in env
+                if for_backward and must_recompute(n) and n in checkpointed_in_bwd:
```

Comment:
maybe I'm missing but although we didn't skip the `n in env` ones when collecting deps,  
due to `n in env`, we always still always skip them in the for loop below.

Reply:
Yeah it is overly confusing. I simplified it